homeassistant/components/mqtt/mixins.py

```diff
@@ -863,12 +875,43 @@ async def async_removed_from_registry(self) -> None:
             self._cleanup_discovery_on_remove()
 
             # Clear the discovery topic so the entity is not rediscovered after a restart
-            await async_remove_discovery_payload(self.hass, self._discovery_data)
+            await async_remove_discovery_payload(
+                self.hass, cast(dict, self._discovery_data)
+            )
+
+    async def _async_clear_discovery_topic_if_entity_removed(
```

Comment:
Either make this an `@callback` if there's a reason for not awaiting `async_remove_discovery_payload`, or await `async_remove_discovery_payload`.